flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/restore/ChangelogBackendLogApplier.java

```diff
@@ -0,0 +1,176 @@
+ * limitations under the License.
+ */
+
+package org.apache.flink.state.changelog.restore;
+
+import org.apache.flink.api.common.state.AggregatingStateDescriptor;
+import org.apache.flink.api.common.state.ListStateDescriptor;
+import org.apache.flink.api.common.state.MapStateDescriptor;
+import org.apache.flink.api.common.state.ReducingStateDescriptor;
+import org.apache.flink.api.common.state.StateDescriptor;
+import org.apache.flink.api.common.state.ValueStateDescriptor;
+import org.apache.flink.api.common.typeutils.base.ListSerializer;
+import org.apache.flink.api.common.typeutils.base.MapSerializer;
+import org.apache.flink.core.memory.DataInputView;
+import org.apache.flink.core.memory.DataInputViewStreamWrapper;
+import org.apache.flink.runtime.state.RegisteredKeyValueStateBackendMetaInfo;
+import org.apache.flink.runtime.state.RegisteredPriorityQueueStateBackendMetaInfo;
+import org.apache.flink.runtime.state.changelog.StateChange;
+import org.apache.flink.runtime.state.metainfo.StateMetaInfoReader;
+import org.apache.flink.runtime.state.metainfo.StateMetaInfoSnapshot;
+import org.apache.flink.runtime.state.metainfo.StateMetaInfoSnapshot.BackendStateType;
+import org.apache.flink.runtime.state.metainfo.StateMetaInfoSnapshotReadersWriters;
+import org.apache.flink.state.changelog.ChangelogKeyedStateBackend;
+import org.apache.flink.state.changelog.ChangelogState;
+import org.apache.flink.state.changelog.StateChangeOperation;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+
+import static org.apache.flink.runtime.state.metainfo.StateMetaInfoSnapshotReadersWriters.StateTypeHint.KEYED_STATE;
+import static org.apache.flink.state.changelog.StateChangeOperation.METADATA;
+import static org.apache.flink.state.changelog.restore.FunctionDelegationHelper.delegateAggregateFunction;
+import static org.apache.flink.state.changelog.restore.FunctionDelegationHelper.delegateReduceFunction;
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/** Applies {@link StateChange}'s to a {@link ChangelogKeyedStateBackend}. */
+@SuppressWarnings({"rawtypes", "unchecked"})
+class ChangelogBackendLogApplier {
+    private static final Logger LOG = LoggerFactory.getLogger(ChangelogBackendLogApplier.class);
+
+    public static void apply(
+            StateChange stateChange,
+            ChangelogKeyedStateBackend<?> changelogBackend,
+            ClassLoader classLoader)
+            throws Exception {
+        DataInputViewStreamWrapper in =
+                new DataInputViewStreamWrapper(new ByteArrayInputStream(stateChange.getChange()));
+        applyOperation(
+                StateChangeOperation.byCode(in.readByte()),
+                stateChange.getKeyGroup(),
+                changelogBackend,
+                in,
+                classLoader,
+                new ChangelogApplierFactoryImpl());
```

Comment:
I think a singleton instance of `ChangelogApplierFactoryImpl` should be better here.

Reply:
I agree, will change it, thanks.